CMakeLists.txt

```diff
@@ -22,6 +22,46 @@ if ( ENABLE_CCACHE )
   set(CMAKE_CXX_COMPILER_LAUNCHER ${CCACHE_PROGRAM})
 endif ()
 
+include(TestBigEndian)
+test_big_endian(BROKER_BIG_ENDIAN)
+
+# Mac OS ignores -pthread but other platforms require it
+if (NOT BROKER_APPLE)
+  set(EXTRA_FLAGS "${EXTRA_FLAGS} -pthread")
+endif ()
+
+# Increase warnings.
+set(EXTRA_FLAGS "${EXTRA_FLAGS} -Wall -Wno-unused -pedantic")
+
+# Increase maximum number of instantiations.
+set(EXTRA_FLAGS "${EXTRA_FLAGS} -ftemplate-depth=512")
+
+option(ENABLE_ADDRESS_SANITIZER "Set to ON to enable the sanitizers")
+
+if ( ENABLE_ADDRESS_SANITIZER )
+  set(additional_flags "-fsanitize=address,undefined -fno-omit-frame-pointer")
```

Comment:
This silently changes the meaning of `ENABLE_ADDRESS_SANITIZER` to a non-obvious behavior. A user that asks for ASAN should not get UBSAN bundled in. Once we add some extra sanitizers by default, where do we draw the line? Some users also might want to enable the thread sanitizer or the memory sanitizer. Eventually, `ENABLE_ADDRESS_SANITIZER` might actually enable all the sanitizers... 😉

I see two approaches work here:
1. Do not touch the `CMakeLists.txt` at all. Users can already set `EXTRA_FLAGS=-fsanitize=address,undefined -fno-omit-frame-pointer` to get the desired outcome without shoe-horning them into exactly ASAN and UBSAN. To make this available to users of the `configure` script, we could instead add a `--extra-flags=` option there. This is the most flexible solution, but certainly requires a bit more "work" on the user side.
2. Drop the `ENABLE_ADDRESS_SANITIZER` flag and instead allow users to list any number of sanitizers (comma separated) in `SANITIZERS`. This is what we do in [`zeek/CMakeLists.txt`](https://github.com/zeek/zeek/blob/master/CMakeLists.txt#L124).

I don't have a strong opinion on which option to pick. @jsiwek do you have a preference between the two?

Reply:
Hello @Neverlord!

I have moved the flag handling so that it happens before the built-in CAF is configured (in order to propagate the settings from the parent project to the child one).

The trouble with EXTRA_FLAGS is that we have to set [both the compiler flags and the linker flags](https://clang.llvm.org/docs/AddressSanitizer.html#usage) according to the documentation. It is not possible to pass the flags through that, as they have to land inside the `CMAKE_EXE_LINKER_FLAGS` (i.e. -Wl,-flag would not work).

Ideally, I think the best approach would be to use CMake options with target-based functions (target_link_options/target_compile_options) and avoid touching CMAKE_CXX_FLAGS/CMAKE_EXE_LINKER_FLAGS directly

Reply:
My thoughts/breakdown:

* For Broker's configuration interface, provide `./configure --sanitizers=` and `BROKER_SANITIZERS` CMake option to users since the `EXTRA_FLAGS` approach might not treat compile versus link flags as precisely we'd need in this case.
* For Broker's implementation of that new flag/option, having it to do the Right thing for both compiler and linker flags sounds fine, but doesn't matter too much to me if it's doing that via `CMAKE_CXX_FLAGS`/`CMAKE_EXE_LINKER_FLAGS` or `target_link_options/target_compile_options`.  The later I guess would be the more Proper/Robust treatment, so feel free to try to get it working via those.
* For the embedded CAF ExternalProject, currently sounds correct to pass sanitizer flags like they are in the PR (`-DEXTRA_FLAGS` and `-DCMAKE_EXE_LINKER_FLAGS`) since `target_compile_options` for `IMPORTED` targets only allows `INTERFACE` items (which I take to mean CAF itself wouldn't be compiled with the desired flags if we went that route)
* For the current meaning of `ENABLE_ADDRESS_SANITIZER`, either don't change it or remove the flag/option entirely since we just provided generalized convenience options for picking sanitizers (I lean more toward the later)

An aside: setting `CXX`/`LD`/`CXXFLAGS`/`LDFLAGS` environment variables is one most-generic way a user can add sanitizer flags (that's originally how I got used to doing it for Zeek), but extending the configure/cmake setup with convenience options does seem helpful.